apps/kbmatry/lib.js

```diff
@@ -0,0 +1,489 @@
+    if (!special) {
+      key.chars = chars[i];
+    }
+    if (key.chars.length > 1) {
+      key.pendingSubKeys = true;
+      key.getSubKeys     = () => getKeys(key.chars);
+      resolve(key)
+    } else {
+      resolve(key);
+    }
+  })
+}
+
+
+/**
+ * Given a set of characters (or sets of characters) get the position and dimensions of the i'th key in that set.
+ * @param charSet An array where each element represents a key on the hypothetical keyboard.
+ * @param i The index of the key in the set you want to get dimensions for.
+ * @param special The special property of the key - for example "del" for a key used for deleting characters.
+ * @returns {{special, bord: number, pad: number, w: number, x: number, h: number, y: number, chars: *[]}}
+ */
+function getKeyByIndex(charSet, i, special) {
+  // Key dimensions
+  const keyboardOffsetY = 40;
+  const margin          = 3;
+  const padding         = 4;
+  const border          = 2;
+  const gridWidth       = Math.ceil(Math.sqrt(charSet.length));
+  const gridHeight      = Math.ceil(charSet.length / gridWidth);
+  const keyWidth        = Math.floor((g.getWidth()) / gridWidth) - margin;
+  const keyHeight       = Math.floor((g.getHeight() - keyboardOffsetY) / gridHeight) - margin;
+  const gridx           = i % gridWidth;
+  const gridy           = Math.floor(i / gridWidth) % gridWidth;
+  const x               = gridx * (keyWidth + margin);
+  const y               = gridy * (keyHeight + margin) + keyboardOffsetY;
+  const w               = keyWidth;
+  const h               = keyHeight;
+  // internal Character spacing
+  const numChars        = charSet[i].length;
+  const subGridWidth    = Math.ceil(Math.sqrt(numChars));
+  const subGridHeight   = Math.ceil(numChars / subGridWidth);
+  const bestFont        = getBestFont(w - padding, h - padding, 0, subGridWidth, subGridHeight);
+  const letterWidth     = bestFont.w;
+  const letterHeight    = bestFont.h;
+  const totalWidth      = (subGridWidth - 1) * (w / subGridWidth) + padding + letterWidth + 1;
+  const totalHeight     = (subGridHeight - 1) * (h / subGridHeight) + padding + letterHeight + 1;
+  const extraPadH       = (w - totalWidth) / 2;
+  const extraPadV       = (h - totalHeight) / 2;
+  return {
+    x,
+    y,
+    w,
+    h,
+    pad  : padding,
+    bord : border,
+    chars: [],
+    special,
+    subGridWidth,
+    subGridHeight,
+    extraPadH,
+    extraPadV,
+    font : bestFont.font
+  };
+}
+
+
+/**
+ * This is probably the most intense part of this keyboard library. If you don't do it ahead of time, it will happen
+ * when you call the keyboard, and it can take up to 1.5 seconds for a full keyboard. Not a super great user experience
+ * SO if you have a tiny keyset, don't worry about it so much, but if you want to maximize performance, generate
+ * a keyboard ahead of time and pass it into the "keyboard" argument of the object in the "input" function.
+ * @param charSets
+ * @returns {Promise<unknown>}
+ */
```

Comment:
Hmm, actually, this comment is out of date. I improved performance significantly since writing this. Will update with more recent performance data.

Reply:
Fixed!